backend/queues/makeOfferProcessor.js

```diff
@@ -34,48 +57,126 @@ async function processor(job) {
+    const wallet = new ethers.Wallet(pk, provider)
+    const walletAddress = wallet.address
+    const marketplace = new ethers.Contract(
+      network.marketplaceContract,
+      marketplaceAbi,
+      wallet
     )
-    queueLog(50, JSON.stringify(tx))
+    queueLog(22, `Using walletAddress ${walletAddress}`)
+
+    let tx, transaction
 
-    // TODO: Code to prevent duplicate txs
-    // TODO Record tx and wait for TX to go through the blockchain
+    // Check if there is a pending transaction for the wallet.
+    // It's possible the processing got interrupted while waiting for the tx to get mined.
+    // For example due to a server maintenance or a crash.
+    transaction = await Transaction.findOne({
+      where: {
+        networkId: network.networkId,
+        wallet: walletAddress,
+        status: TransactionStatuses.Pending
+      }
+    })
+    if (transaction) {
+      log.info(
+        `Found pending transaction ${transaction.id} job ${job.id} hash ${transaction.hash} for wallet ${walletAddress}`
+      )
+
+      // If it is not the transaction from our job. Do not try to recover it.
+      // Let it get recovered by the job that created it.
+      // Fail our job for now, it will get retried.
+      if (transaction.jobId === job.id) {
+        throw new Error(
+          `Pending transaction does not belongs to job ${job.id}. Bailing.`
+        )
+      }
+
+      // Try to recover by loading the tx from its hash.
+      tx = await provider.getTransaction(transaction.hash)
+      if (!tx) {
+        // The transaction was not mined and is not in the transaction pool.
+        // Something went really wrong...
+        throw new Error(
+          `Transaction ${transaction.id} with hash ${transaction.hash} not found`
+        )
+      }
+      log.info('Recovered tx', tx)
+    } else {
+      // Send a blockchain transaction to make an offer on the marketplace contract.
+      queueLog(25, 'Sending to marketplace')
+      tx = await _createOffer(
+        provider,
+        marketplace,
+        wallet,
+        lid,
+        offer,
+        ipfsHash
+      )
+      log.info('Transaction sent:', tx)
+
+      // Record the transaction in the DB.
+      transaction = await Transaction.create({
+        shopId,
+        networkId: network.networkId,
+        wallet: walletAddress,
+        type: TransactionTypes.OfferCreated,
+        status: TransactionStatuses.Pending,
+        hash: tx.hash,
+        listingId: lid.toString(),
+        ipfsHash,
+        jobId: job.id.toString()
+      })
+    }
+
+    // Wait for the tx to get mined.
+    // Note: this is blocking with no timeout. Depending on the network conditions,
+    // it could take a long time for a tx to get mined. This will become a bottleneck
+    // in the future when the transaction volume scales up. A potential solution
+    // will be to run the confirmation logic as a separate queue with multiple workers.
```

Comment:
FYI: I think there will be multiple workers.  Right now, should be 3 for prod.

Reply:
This is predicated on a single worker processing the queue.
Afaik by default Bull's concurrency is set to 1: https://github.com/OptimalBits/bull/blob/develop/REFERENCE.md#queueprocess

Added a comment to make sure concurrency does not get increased inadvertently in the future.

Reply:
I think that concurrency setting only accounts for workers in that process, not multiple processes.  We currently have 3 instances in prod.  It don't think it will run the same job concurrently, but I'm not sure what you're concerned about, either.  Just wanted to point out that blocking here doesn't necessarily hold up other jobs like your comment suggested.

Reply:
I'm so glad we do code reviews! THANK YOU. I was under the incorrect assumption that Bull's concurrency settings are global across node.js processes. They are not. Let me rework a bit the code. I think I can still make it work by checking if there is any pending transaction from the wallet address and if any and it's not the same job id then retry...

Reply:
Inspected the logic and it should work for multi-consumer nodes.